### ISSUE TITLE
Lint suggestion: `assume_init_mut` passed to `drop_in_place`

### ISSUE THREAD

**Issue by clarfonthey** [rust-lang/rust-clippy#13860]
### What it does

Since I found a few of these while working on rust-lang/rust#129259, this feels like a decent lint.

Assuming that this will be worked on after that PR is merged, it should detect both `MaybeUninit::assume_init_mut` and `<[MaybeUninit<_>]>::assume_init_mut`.

### Advantage

* Removes need to import `drop_in_place` in many cases
* Fewer method calls = better codegen
* Probably easier to read due to being shorter

### Drawbacks

It could be difficult to detect all cases, e.g. binding the `assume_init_mut` value to a variable and then calling `drop_in_place`.

### Example

Essentially, it would look for calls of the form:

```rust
ptr::drop_in_place(x.assume_init_mut())
```

and replace them with:

```rust
x.assume_init_drop()
```

We probably also want to cover the simple case of:

```rust
let y = x.assume_init_mut();
ptr::drop_in_place(y);
```

And… well, if `y` is used after that, this would be UB anyway, so, I'm going to assume that it's okay to replace this with:

```rust
x.assume_init_drop()
```
